File: harmony-app/entry/src/main/ets/model/ApiClient.ts
import http from '@ohos.net.http'
import { Todo, DeviceInfo } from '../model/Todo'

export class ApiClient {
  private baseURL: string
  private token: string = ''
  private httpClient: http.HttpRequest

  constructor(baseURL: string) {
    this.baseURL = baseURL
    this.httpClient = http.createHttp()
  }

  setToken(token: string) {
    this.token = token
  }

  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json'
    }
    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`
    }
    return headers
  }

  async register(): Promise<DeviceInfo> {
    const resp = await this.httpClient.request(
      `${this.baseURL}/api/register`,
      {
        method: http.RequestMethod.POST,
        extraData: JSON.stringify({ source: 'harmony' }),
        header: this.buildHeaders()
      }
    )
    const body = JSON.parse(resp.result as string)
    return DeviceInfo.fromJson(body)
  }

  async fetchTodos(since: string): Promise<Todo[]> {
    const resp = await this.httpClient.request(
      `${this.baseURL}/api/todos?since=${since}`,
      {
        method: http.RequestMethod.GET,
        header: this.buildHeaders()
      }
    )
    const body = JSON.parse(resp.result as string) as Record<string, Object>[]
    return body.map(item => Todo.fromJson(item))
  }

  async createTodo(todo: Todo): Promise<Todo> {
    const resp = await this.httpClient.request(
      `${this.baseURL}/api/todos`,
      {
        method: http.RequestMethod.POST,
        extraData: JSON.stringify({
          title: todo.title, notes: todo.notes,
          dueDate: todo.dueDate, priority: todo.priority,
          isCompleted: todo.isCompleted, source: 'harmony'
        }),
        header: this.buildHeaders()
      }
    )
    const body = JSON.parse(resp.result as string)
    return Todo.fromJson(body)
  }

  async updateTodo(id: string, fields: Record<string, Object>): Promise<Todo> {
    const resp = await this.httpClient.request(
      `${this.baseURL}/api/todos/${id}`,
      {
        method: http.RequestMethod.PUT,
        extraData: JSON.stringify({ fields }),
        header: this.buildHeaders()
      }
    )
    const body = JSON.parse(resp.result as string)
    return Todo.fromJson(body)
  }

  async deleteTodo(id: string): Promise<void> {
    await this.httpClient.request(
      `${this.baseURL}/api/todos/${id}`,
      {
        method: http.RequestMethod.DELETE,
        header: this.buildHeaders()
      }
    )
  }
}

File: macos-app/TodoSync/Sources/TodoSync/ApiClient.swift
import Foundation

@MainActor
class ApiClient {
    let baseURL: String
    var token: String?
    let session = URLSession.shared

    init(baseURL: String) {
        self.baseURL = baseURL
    }

    func register() async throws -> DeviceInfo {
        let url = URL(string: "\(baseURL)/api/register")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["source": "macos"])
        let (data, _) = try await session.data(for: req)
        return try JSONDecoder().decode(DeviceInfo.self, from: data)
    }

    func fetchTodos(since: String) async throws -> [Todo] {
        let url = URL(string: "\(baseURL)/api/todos?since=\(since)")!
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token ?? "")", forHTTPHeaderField: "Authorization")
        let (data, _) = try await session.data(for: req)
        return try JSONDecoder().decode([Todo].self, from: data)
    }

    func createTodo(_ todo: Todo) async throws -> Todo {
        let url = URL(string: "\(baseURL)/api/todos")!
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token ?? "")", forHTTPHeaderField: "Authorization")
        let body: [String: Any] = [
            "title": todo.title, "notes": todo.notes,
            "dueDate": todo.dueDate ?? "", "priority": todo.priority,
            "isCompleted": todo.isCompleted, "source": "macos"
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await session.data(for: req)
        return try JSONDecoder().decode(Todo.self, from: data)
    }

    func updateTodo(id: String, fields: [String: [String: Any]]) async throws -> Todo {
        let url = URL(string: "\(baseURL)/api/todos/\(id)")!
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token ?? "")", forHTTPHeaderField: "Authorization")
        req.httpBody = try JSONSerialization.data(withJSONObject: ["fields": fields])
        let (data, _) = try await session.data(for: req)
        return try JSONDecoder().decode(Todo.self, from: data)
    }

    func deleteTodo(id: String) async throws {
        let url = URL(string: "\(baseURL)/api/todos/\(id)")!
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        req.setValue("Bearer \(token ?? "")", forHTTPHeaderField: "Authorization")
        let _ = try await session.data(for: req)
    }

    // Called when user adds/edits reminders directly in Reminders.app
    func pushLocalChanges(todos: [Todo], deviceId: String) async throws {
        for todo in todos {
            var fields: [String: [String: Any]] = [:]
            let now = ISO8601DateFormatter().string(from: Date())
            fields["title"] = ["value": todo.title, "updatedAt": now, "deviceId": deviceId]
            fields["notes"] = ["value": todo.notes, "updatedAt": now, "deviceId": deviceId]
            fields["isCompleted"] = ["value": todo.isCompleted, "updatedAt": now, "deviceId": deviceId]
            _ = try await updateTodo(id: todo.id, fields: fields)
        }
    }
}
